src/translators/typescript.py

```diff
@@ -0,0 +1,748 @@
+        if getattr(node.class_type, 'can_infer_type_args', None) is True:
+            self._children_res.append("new {}({})".format(
+                " " * self.ident + node.class_type.name,
+                ", ".join(children_res)))
+        else:
+            self._children_res.append("new {}({})".format(
+                " " * self.ident + self.get_type_name(node.class_type),
+                ", ".join(children_res)))
+
+    @append_to
+    def visit_field_access(self, node):
+        old_ident = self.ident
+        self.ident = 0
+        children = node.children()
+        for c in children:
+            c.accept(self)
+        children_res = self.pop_children_res(children)
+        self.ident = old_ident
+        receiver_expr = (children_res[0] if children_res[0]
+                        else "this")
+        res = "{}.{}".format(receiver_expr, node.field)
+        self._children_res.append(res)
+
+    @append_to
+    def visit_func_ref(self, node):
+        def needs_this_prefix(node, decl):
+            if node.receiver is not None:
+                return False
+            elif decl is None:
+                return True # Function is an inherited method
+            elif isinstance(decl, ast.FunctionDeclaration) and decl.is_class_method():
+                return True # Function is method of current class
+            elif isinstance(decl, ast.FieldDeclaration) and decl.get_type().name[:-1] == tst.TypeScriptBuiltinFactory().get_function_type().name[:-1]:
+                return True # Function is callable field
+            return False
+
+        old_ident = self.ident
+
+        self.ident = 0
+        children = node.children()
+        for c in children:
+            c.accept(self)
+
+        self.ident = old_ident
+        children_res = self.pop_children_res(children)
+
+        this_prefix = children_res[0] if children_res else ""
+
+        decl = get_decl(self.context, self._namespace, node.func)
+
+        if decl is not None:
+            _, decl = decl
+
+        if needs_this_prefix(node, decl):
+            this_prefix += "this"
+            # FIXME Must check signatures and not names
+            # (for overwritten + overloaded functions)
+
+        res = "{}{}{}".format(" "*self.ident, "({}).".format(this_prefix) if this_prefix else "", node.func)
+
+        self._children_res.append(res)
+
+    @append_to
+    def visit_func_call(self, node):
+        def needs_this_prefix(node, decl):
+            if node.receiver is not None:
+                return False
+
+            elif decl is None: # Function is an inherited method
+                return True
+            elif isinstance(decl, ast.FunctionDeclaration) and decl.is_class_method(): # Function is method of current class
+                return True
+            elif isinstance(decl, ast.FieldDeclaration) and decl.get_type().name[:-1] == tst.TypeScriptBuiltinFactory().get_function_type().name[:-1]: # Function is callable field
```

Comment:
the same here. this line is too long.

Also, I noticed that the method `needs_this_prefix` is used in two different places. So, I think we should extract a top-level function from this nested function.